x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/dataframe/transforms/DataFrameTransformCheckpoint.java

```diff
@@ -0,0 +1,268 @@
+        parser.declareLong(constructorArg(), TIMESTAMP);
+        parser.declareLong(constructorArg(), CHECKPOINT);
+
+        parser.declareObject(constructorArg(), (p,c) -> {
+            Map<String, long[]> checkPointsByIndexName = new TreeMap<>();
+            XContentParser.Token token = null;
+            while ((token = p.nextToken()) != XContentParser.Token.END_OBJECT) {
+                if (token != XContentParser.Token.FIELD_NAME) {
+                    throw new ParsingException(p.getTokenLocation(), "Unexpected token " + token + " ");
+                }
+
+                final String indexName = p.currentName();
+                token = p.nextToken();
+                if (token != XContentParser.Token.START_ARRAY) {
+                    throw new ParsingException(p.getTokenLocation(), "Unexpected token " + token + " ");
+                }
+
+                long[] checkpoints = p.listOrderedMap().stream().mapToLong(num -> ((Number) num).longValue()).toArray();
+                checkPointsByIndexName.put(indexName, checkpoints);
+            }
+            return checkPointsByIndexName;
+        }, INDICES);
+        parser.declareLong(optionalConstructorArg(), TIME_UPPER_BOUND);
+        parser.declareString(optionalConstructorArg(), DataFrameField.INDEX_DOC_TYPE);
+
+        return parser;
+    }
+
+    public DataFrameTransformCheckpoint(String id, Long timestamp, Long checkpoint, Map<String, long[]> checkpoints, Long timeUpperBound) {
+        this.id = id;
+        this.timestampMillis = timestamp.longValue();
+        this.checkpoint = checkpoint;
+        this.indicesCheckpoints = checkpoints;
+        this.timeUpperBoundMillis = timeUpperBound == null ? 0 : timeUpperBound.longValue();
+    }
+
+    public DataFrameTransformCheckpoint(StreamInput in) throws IOException {
+        this.id = in.readString();
+        this.timestampMillis = in.readLong();
+        this.checkpoint = in.readLong();
+        this.indicesCheckpoints = readCheckpoints(in.readMap());
+        this.timeUpperBoundMillis = in.readLong();
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        final String humanReadableSuffix = "_string";
+        builder.startObject();
+
+        // the id, doc_type and checkpoint is only internally used for storage, the user-facing version gets embedded
+        if (params.paramAsBoolean(DataFrameField.FOR_INTERNAL_STORAGE, false)) {
+            builder.field(DataFrameField.ID.getPreferredName(), id);
+            builder.field(CHECKPOINT.getPreferredName(), checkpoint);
+            builder.field(DataFrameField.INDEX_DOC_TYPE.getPreferredName(), NAME);
+        }
+
+        builder.field(TIMESTAMP.getPreferredName(), timestampMillis);
+        if (timeUpperBoundMillis > 0) {
+            builder.field(TIME_UPPER_BOUND.getPreferredName(), timeUpperBoundMillis);
+        }
+
+        if (params.paramAsBoolean("human", false)) {
```

Comment:
Use `builder.timeField(String name, String readableName, long value)` to get the human readable dates